server/src/main/java/org/opensearch/ingest/IngestService.java

```diff
@@ -211,11 +211,18 @@ public static boolean resolvePipelines(
                     finalPipeline = IndexSettings.FINAL_PIPELINE.get(indexSettings);
                     indexRequest.setFinalPipeline(finalPipeline);
                 }
-            } else if (indexRequest.index() != null) {
+            } else if (indexRequest.index() != null || originalRequest != null && originalRequest.index() != null) {
```

Comment:
I think maybe it's by design that `index` field is missing in both `doc` and `upsert` because of this line:
https://github.com/opensearch-project/OpenSearch/blob/ba20bdae4db6700e0bc530d339197dd2942a401f/server/src/main/java/org/opensearch/ingest/IngestService.java#L196
, but your solution also works, I've tried that and run all tests, no error found for now.